report/3.Theory/ParserImplementation.tex

```diff
@@ -0,0 +1,96 @@
+\section{Parser implementation}
+
+Since the compiler language kotlin is completely interoperable with java, both SableCC and ANLTR could be obvious choices for parser if the group decided to go with a generated parser. 
+However, for the parser of HCL it was decided that since HCL's CFG is LL(k), the parser should be handwritten. 
+This is partly due to the control that a handwritten parser offers, but primarily out of curiosity and for educational purposes.
+
+Due to the nature of HCL, the parser can be implemented as a single pass parser. 
+This is because the user may not access any variables that has not yet been declared. 
+Furthermore as mentioned HCL is a LL(K) language, and it was decided that the parser should be a LL(K) recursive decent parser.
+
+Both syntactical and contextual analysis is done within the parser. 
+This means that there is no additional post parser type checking or contextual analysis step. 
+One major reason for doing in line contextual analysis is the way functions are invoked in HCL. 
+The only way to know whether anything is a function call, is by analyzing the types. 
+This means that in order to create a proper abstract syntax tree the type checking must be done in line. 
+It could also potentially be done afterwards, but that would require quite a big amount of post processing, that would be a lot more complex, than just doing type checking within the parser. 
+This also means that the outputted abstract syntax tree is in fact already a decorated syntax tree. 
+For instance function calls already has information on its return type, even though the function may have multiple return types based on the overloading of the function.
+
+\subsection{Implementing a recursive decent parser}
+
+A LL(K) recursive decent parser will do look aheads until it is aware what branch should be parsed.
+This can be figured out from the first and follow sets, that can be produced from the CFG.
+Once it figures out which non-terminal to parse it will call a parse method that should parse that part.
+The parse method can then accept terminals or parse further non-terminals.
+It should be the responsibility of the accept method to advance the lexical token stream.
+
+Lets have a quick and concrete example from HCL. 
```

Comment:
this sentence turns the report into a follow-along adventure for a kids cartoon show.
"below is a quick.."